retrievalmarket/types.go

```diff
@@ -416,3 +416,21 @@ type ChannelAvailableFunds struct {
 	// and in the local datastore
 	VoucherReedeemedAmt abi.TokenAmount
 }
+
+// PricingInput provides input parameters required to price a retrieval deal.
+type PricingInput struct {
+	// PayloadCID is the cid of the payload to retrieve.
+	PayloadCID cid.Cid
+	// PieceCID is the cid of the Piece from which the Payload will be retrieved.
+	PieceCID cid.Cid
+	// PieceSize is the size of the Piece from which the payload will be retrieved.
+	PieceSize abi.UnpaddedPieceSize
+	// Client is the peerID of the retrieval client.
+	Client peer.ID
+	// VerifiedDeal is true if there exists a verified storage deal for the PayloadCID.
+	VerifiedDeal bool
+	// Unsealed is true if there exists an unsealed sector from which we can retrieve the given payload.
+	Unsealed bool
+	// CurrentAsk is the current configured ask in the ask-store.
+	CurrentAsk Ask
+}
```

Comment:
( I am well aware I am commenting on an old, merged PR )

@aarshkshah1992 @hannahhoward @willscott was it a deliberate decision to exclude the client-provided selector in this structure, or just an omission? Even if we do not have the tools for miners to interpret/decide on a selector-driven partial-retrieval, the structure is [being exposed to 3rd party tools](https://github.com/filecoin-project/lotus/discussions/6780#discussion-3466055). It is probably prudent to provide all the information available from the get-go, instead of adding it later on.

Reply:
An extra nit regarding the title-comment:

> If the Miner has multiple Sector files containing the required retrieval payload AND the client hasn't mandated that we retrieve the payload from a specific Piece,

A specific deal `PayloadCID` (root-dag) + `PieceSize` will always invariably correspond to an identical `PieceCID`. That is - if a deal is stored in multiple sectors, some sealed and some unsealed, a `PieceCID` can not differentiate among them: it will be identical among them all.

Reply:
Never too late to provide useful feedback! ;-)

Re: including the selector in the pricing function input: how useful would this be? The function can't _evaluate_ the selector against the data, and it can't really derive anything useful from it intrinsically AFAIK.

If it's to symbolise that a partial retrieval is being performed, for pricing purposes, the price-per-byte formulation of the quote already decouples the final price from the actual data being selected. We _could_ pass a "partial retrieval = true/false" flag, but it's not possible to calculate it correctly because the selector could be selecting the entire deal.

Re: extensibility/completeness, it's uninvasive to add a JSON field in the future if we do decide we want this for some reason. But until there's a standardised, canonical text representation for selectors, I don't think it's practical to even consider.

Reply:
> A specific deal PayloadCID (root-dag) + PieceSize will always invariably correspond to an identical PieceCID. That is - if a deal is stored in multiple sectors, some sealed and some unsealed, a PieceCID can not differentiate among them: it will be identical among them all.

If there's an unsealed copy, we always select the unsealed copy as input to the pricing function, and as the source of the retrieval.

Reply:
@raulk right, I was specifically questioning this part of the sentence:

> the client hasn't mandated that we retrieve the payload from a specific Piece

There is no way for a client to signal "I want to retrieve from this physical piece" via the struct in question: they all have the same `PieceCid`

Reply:
Oh, what @aarshkshah1992 was trying to get to is that if "universal retrieval" is enabled (such that clients don't need to specify the PieceCID they want to retrieval a given PayloadCID from), and there are multiple PieceCIDs that contain that PayloadCID, we prefer the deal that's unsealed, should there be one. I don't think any of that implies that the client gets to choose the sector.

Reply:
> But until there's a standardized, canonical text representation for selectors, I don't think it's practical to even consider.

The reason I brought this up is that the selector is already exposed in a [standard-ish encoded-cbor representation](https://github.com/filecoin-project/go-fil-markets/blob/v1.6.0-rc1/retrievalmarket/types.go#L245-L253) via the `RetrievalFilter` codepath: https://github.com/filecoin-project/lotus/blob/v1.11.0-rc2/markets/dealfilter/cli.go#L28-L62

I do not have a preference either way. Rather am raising the inconsistency of having two sets of external tools, one being fed the structure in this pr, the other being fed this: https://github.com/Murmuration-Labs/bitscreen/blob/master/testdata/retrieval_proposal.json

Reply:
Thanks for the pointer. Note that this isn't a filtering function, but rather a pricing function. It runs when the retrieval provider is _asked_ for a quote. At that time, the selector is not available. So in terms of the data interface, the deal filter and the pricing function can't be symmetrical.

Reply:
> It runs when the retrieval provider is asked for a quote. At that time, the selector is not available.

@raulk indeed you are correct: it is all commented out :cryingbear:
https://github.com/filecoin-project/go-fil-markets/blob/v1.6.0-rc1/retrievalmarket/types.go#L158-L167